rfcs/text/0014_api_documentation.md

```diff
@@ -0,0 +1,439 @@
+
+## Post MVP
+
+- Plugin `{@link AnApi}` links work. Will need to decide if we only support per plugin links, or if we should support a way to do this across plugins.
+- Ingesting stats like number of public APIs, and number of those missing comments
+- Include and expose API references
+- Use namespaces to split large plugins
+
+# Information available for each API declaration
+
+We have the following pieces of information available from each declaration:
+
+- Label. The name of the function, class, interface, etc.
+
+- Description. Any comment that was able to be extracted. Currently it's not possible for this data to be formatted, for example if it has a code example with back tics. This
+is dependent on the elastic-docs team moving the infrastructure to NextJS instead of Gatsby, but it will eventually be supported.
+
+- Tags. Any `@blahblah` tags that were extracted from comments. Known tags, like `beta`, will be show help text in a tooltip when hovered over.
+
+- Type. This can be thought of as the _kind_ of type (see [TypeKind](#typekind)). It allows us to group each type into a category. It can be a primitive, or a
+more complex grouping. Possibilities are: array, string, number, boolean, object, class, interface, function, compound (unions or intersections)
+
+- Required or optional. (whether or not the type was written with `| undefined` or `?`). This terminology makes the most sense for function
+parameters, not as much when thinking about an exported variable that might be undefined.
+
+- Signature. This is only relevant for some types: functions, objects, type, arrays and compound. Classes and interfaces would be too large.
+For primitives, this is equivalent to "type".
+
+- Children. Only relevant for some types, this would include parameters for functions, class members and functions for classes, properties for
+interfaces and objects. This makes the structure recursive. Each child is a nested API component.
+
+- Return comment. Only relevant for function types.
+
+![image](../images/api_info.png)
+
+
+### ApiDeclaration type
+
+```ts
+interface ApiDeclaration {
+  label: string;
+  type: TypeKind; // string, number, boolean, class, interface, function, type, etc.
+  description: TextWithLinks;
+  signature: TextWithLinks;
+  tags: string[];  // Declarations may be tagged as beta, or deprecated.
+  children: ApiDeclaration[]; // Recursive - this could be function parameters, class members, or interface/object properties.
+  returnComment?: TextWithLinks;
+  lifecycle?: Lifecycle.START | Lifecycle.SETUP;
+}
+
+```
+
+# Architecture design
+
+## Location
+
+The generated docs will reside inside the kibana repo, inside a top level `api_docs` folder. In the long term, we could investigate having the docs system run a script to generated the mdx files, so we don’t need to store them inside the repo. Every ci run should destroy and re-create this folder so removed plugins don't have lingering documentation files.
```

Comment:
I don't think so - I always found it really weird that we have `docs/developer` and `docs/development` and one is for autotgenerated md files and the other is hand written asciidoc. 

If we could split `docs` into two sub folders  (one hand written, the other auto generated) I think that would be okay, but that would be a lot work. Note we also now have `dev_docs` which I'm using for high level, cross plugin documentation (hand written).

Reply:
> I don't think so - I always found it really weird that we have docs/developer and docs/development and one is for autotgenerated md files and the other is hand written asciidoc.

Agreed that the current folder-structure is super awkward.

> If we could split docs into two sub folders (one hand written, the other auto generated) I think that would be okay, but that would be a lot work. Note we also now have dev_docs which I'm using for high level, cross plugin documentation (hand written).

Maybe we leave `docs` as the hand-written docs, and add a `generated_docs` folder with `plugin_api` as the child? I'm primarily trying to prevent us from continuing to add `${something}_docs` folders because this will soon become unwieldy.

Reply:
Resolved offline. I'll continue with the current approach, but will restructure if any more top level doc folders show up, or once the current `docs` folder no longer contains asciidoc.